[CALCITE-6388] PsTableFunction throws NumberFormatException when the …

asolimando opened the pull request:
…'user' column has spaces

snuyanzin commented:
I guess the problem here that other columns also could contain space separated values like e.g. `comm`

I tried this `./sqlsh select distinct \`user\` from ps` for the branch for this PR
and the output is
```
6.9
13.3
0.0
0.1
1.0
0.2
1.1
0.3
1.2
69.7
39.5
0.4
0.5
0.6
0.7
108
4.4
0.8
1.8
35.8
0.9

```
while for main same query is
```
message+
serg
avahi
lp
kernoops
root
rtkit
syslog
systemd+
colord
```
after some checks I noticed that `command` could also contain space separated values and probably this is the reason...

caicancai reviewed (commented):
Can we add some tests to illustrate the jira case?

asolimando commented:
@snuyanzin, you are right, both `user` and `comm` can contain spaces, I have looked up if any of the other fields could be used in an unambiguous manner as an anchor to understand when `user` is over, unfortunately it doesn't seem to be any.

What I came up with is to move `user` as first field, split by whitespace as before, and stop parsing `user` whenever a purely numeric token is found. For `comm` I am just composing all the tokens counting on the fact that the number of fields is known.

So at the moment we support spaces, but not when one of the tokens is purely numeric like `root 123`. I don't have better ideas, it seems already an improvement over the current implementation which doesn't properly support spaces neither in `user` nor in `comm`.

@caicancai I have added a test, I had to unpack a bit the current implementation to that aim, thanks for the reminder.

Eager to hear your thoughts, thanks for your comments!

asolimando commented:
> @snuyanzin, you are right, both `user` and `comm` can contain spaces, I have looked up if any of the other fields could be used in an unambiguous manner as an anchor to understand when `user` is over, unfortunately it doesn't seem to be any.
> 
> What I came up with is to move `user` as first field, split by whitespace as before, and stop parsing `user` whenever a purely numeric token is found. For `comm` I am just composing all the tokens counting on the fact that the number of fields is known.
> 
> So at the moment we support spaces, but not when one of the tokens is purely numeric like `root 123`. I don't have better ideas, it seems already an improvement over the current implementation which doesn't properly support spaces neither in `user` nor in `comm`.
> 
> @caicancai I have added a test, I had to unpack a bit the current implementation to that aim, thanks for the reminder.
> 
> Eager to hear your thoughts, thanks for your comments!

Hello, if you don't have further comments I will merge around the end of the week

caicancai commented:
I will review this PR as soon as possible in the next two days

caicancai commented on plus/src/test/java/org/apache/calcite/adapter/os/PsTableFunctionTest.java:
```diff
@@ -0,0 +1,69 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Unit tests for the ps (process status) table function.
```
Add jira link

caicancai commented on plus/src/main/java/org/apache/calcite/adapter/os/PsTableFunction.java:
```diff
@@ -43,114 +46,185 @@ public class PsTableFunction {
+  @VisibleForTesting
+  protected static class LineParser implements Function1<String, Object[]> {
+
+    @Override public Object[] apply(String line) {
```
Good job

caicancai commented on plus/src/main/java/org/apache/calcite/adapter/os/PsTableFunction.java:
```diff
@@ -43,114 +46,185 @@ public class PsTableFunction {
+      } else {
+        values[fieldIdx] = field(PS_FIELD_NAMES.get(fieldIdx), tokens[processedTokens]);
+      }
+
```
Remove extra blank lines

caicancai reviewed (approved):
overall LGTM，I left a few small comments

asolimando commented:
The force push is just to change the latest commit message (forgot that it must now start with an uppercase letter), no impact
